contracts/contracts/chains/polygon_zkevm/PolygonZkEVMMessengers.sol

```diff
@@ -0,0 +1,59 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.19;
+
+import "../../../interfaces/IMessenger.sol";
+import "../../RestrictedCalls.sol";
+import "./PolygonZkBridgeInterface.sol";
+
+// This messenger is used by both L1 & L2 messengers.
+// local means it is on the same chain and remote means
+// it is on opposite chain.
+contract PolygonZkEVMMessenger is IMessenger, RestrictedCalls {
+    IPolygonZkEVMBridge public immutable bridge;
+    address public immutable localCallee;
+    address public remoteMessenger;
+    // 1 means Polygon and 0 means L1
+    uint32 public immutable remoteNetwork;
+
+    constructor(address _bridge, address _localCallee, uint32 _remoteNetwork) {
+        bridge = IPolygonZkEVMBridge(_bridge);
+        localCallee = _localCallee;
+        remoteNetwork = _remoteNetwork;
+    }
+
+    function setRemoteMessenger(address _remoteMessenger) public onlyOwner {
+        require(_remoteMessenger == address(0), "Resolver already set");
+        remoteMessenger = _remoteMessenger;
+    }
+
+    // Messenger is the direct courier for the localCallee because
+    // this contract calls localCallee in onMessageReceived.
+    // We dont check for caller because we already do that onMessageReceived.
+    function callAllowed(
+        address _caller,
+        address courier
+    ) external view returns (bool) {
+        return courier == address(this);
+    }
+
+    // This function is callback for claimMessage and is receiving message
```

Comment:
`is the callback and is receiving the message from the native bridge. The origin address..`